sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -3209,6 +3222,19 @@ class Dataset[T] private[sql](
     }
   }
 
+  private[sql] def getRowsToPython(
+      _numRows: Int,
+      truncate: Int,
+      vertical: Boolean): Array[Any] = {
+    EvaluatePython.registerPicklers()
+    val numRows = _numRows.max(0).min(Int.MaxValue - 1)
+    val rows = getRows(numRows, truncate, vertical).map(_.toArray).toArray
+    val toJava: (Any) => Any = EvaluatePython.toJava(_, ArrayType(ArrayType(StringType)))
+    val iter: Iterator[Array[Byte]] = new SerDeUtil.AutoBatchedPickler(
+      rows.iterator.map(toJava))
+    PythonRDD.serveIterator(iter, "serve-GetRows")
```

Comment:
`PythonRDD.serveIterator(iter, "serve-GetRows")` returns `Int`, but the return type of `getRowsToPython `  is `Array[Any]`. How does it work? cc @xuanyuanking @HyukjinKwon

Reply:
I think we return `Array[Any]` for `PythonRDD.serveIterator` too.

https://github.com/apache/spark/blob/628c7b517969c4a7ccb26ea67ab3dd61266073ca/core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala#L400

Did I maybe miss something?

Reply:
Same answer with @HyukjinKwon about the return type, and actually the exact return type we need here is Array[Array[String]], this defined in `toJava` func.

Reply:
The changes of the return types were made in a commit `[PYSPARK] Update py4j to version 0.10.7.`... No PR was opened...

Reply:
Yup ..

Reply:
re the py4j commit - there's a good reason for it @gatorsmile 
not sure if the change to return type is required with the py4j change though

Reply:
I think we better don't talk about this here though.